### ISSUE TITLE
raise socket.error("socket connection broken")

### ISSUE THREAD

**Issue by YurieCo** [AirtestProject/Airtest#167]
**Describe the bug**
My problem is exact the "https://www.bountysource.com/issues/61631149-socket-connection-broken"
how I can solve it ?

I am trying to mirror a screen and another thread is doing unlock and bot actions like click on something do something. Whats wrong ?

Second thread use device.shell  <-- to grab xml then use device.swipe ?
```
....\venv\Scripts\python.exe C:/Users/pgi/Downloads/adbmirror/adbmirror/gui/filebar/pr.py
[06:28:53][DEBUG]<airtest.core.android.adb> ....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe devices
DEBUG:airtest.core.android.adb:....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe devices
[06:28:53][DEBUG]<airtest.core.android.adb> ....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe devices
DEBUG:airtest.core.android.adb:....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe devices
[06:28:53][DEBUG]<airtest.core.android.adb> ....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF wait-for-device
DEBUG:airtest.core.android.adb:....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF wait-for-device
[06:28:53][DEBUG]<airtest.core.android.adb> ....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF shell getprop ro.build.version.sdk
DEBUG:airtest.core.android.adb:....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF shell getprop ro.build.version.sdk
[06:28:53][DEBUG]<airtest.core.android.adb> ....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF shell ls /data/local/tmp/minicap ; echo ---$?---
DEBUG:airtest.core.android.adb:....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF shell ls /data/local/tmp/minicap ; echo ---$?---
[06:28:53][DEBUG]<airtest.core.android.adb> ....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF shell dumpsys window policy ; echo ---$?---
DEBUG:airtest.core.android.adb:....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF shell dumpsys window policy ; echo ---$?---
[06:28:54][DEBUG]<airtest.core.android.adb> ....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF shell ls /data/local/tmp/minicap.so ; echo ---$?---
DEBUG:airtest.core.android.adb:....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF shell ls /data/local/tmp/minicap.so ; echo ---$?---
[06:28:54][DEBUG]<airtest.core.android.adb> ....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF shell input keyevent POWER ; echo ---$?---
DEBUG:airtest.core.android.adb:....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF shell input keyevent POWER ; echo ---$?---
[06:28:55][DEBUG]<airtest.core.android.adb> ....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF shell LD_LIBRARY_PATH=/data/local/tmp /data/local/tmp/minicap -v 2>&1
DEBUG:airtest.core.android.adb:....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF shell LD_LIBRARY_PATH=/data/local/tmp /data/local/tmp/minicap -v 2>&1
[06:28:55][DEBUG]<airtest.core.android.adb> ....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF shell uiautomator dump && cat /sdcard/window_dump.xml ; echo ---$?---
DEBUG:airtest.core.android.adb:....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF shell uiautomator dump && cat /sdcard/window_dump.xml ; echo ---$?---
[06:28:58][DEBUG]<airtest.core.android.minicap> version:5
DEBUG:airtest.core.android.minicap:version:5
[06:28:58][DEBUG]<airtest.core.android.minicap> skip install minicap
DEBUG:airtest.core.android.minicap:skip install minicap
[06:28:58][DEBUG]<airtest.core.android.adb> ....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF forward --no-rebind tcp:13652 localabstract:minicap_13652
DEBUG:airtest.core.android.adb:....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF forward --no-rebind tcp:13652 localabstract:minicap_13652
<hierarchy rotation="0"><node index="0" text="" resource-id="" class="android.widget.FrameLayout" package="com.android.keyguard" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,0][480,854]"><node index="0" text="" resource-id="com.android.keyguard:id/keyguard_host_view" class="android.widget.FrameLayout" package="com.android.keyguard" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,0][480,854]"><node index="0" text="" resource-id="com.android.keyguard:id/sliding_layout" class="android.view.View" package="com.android.keyguard" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,0][480,854]"><node index="0" text="" resource-id="" class="android.widget.FrameLayout" package="com.android.keyguard" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,0][480,854]" /><node index="2" text="" resource-id="" class="android.widget.FrameLayout" package="com.android.keyguard" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,38][480,854]" /><node index="3" text="" resource-id="" class="android.widget.FrameLayout" package="com.android.keyguard" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,38][480,854]"><node index="0" text="" resource-id="com.android.keyguard:id/app_widget_container" class="android.view.View" package="com.android.keyguard" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="true" long-clickable="false" password="false" selected="false" bounds="[0,0][480,854]"><node index="0" text="" resource-id="com.android.keyguard:id/keyguard_status_view" class="android.widget.FrameLayout" package="com.android.keyguard" content-desc="Status widget." checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="true" password="false" selected="false" bounds="[38,76][442,758]"><node index="0" text="" resource-id="com.android.keyguard:id/keyguard_status_view_face_palm" class="android.widget.GridLayout" package="com.android.keyguard" content-desc="Status" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[50,88][430,344]"><node index="0" text="" resource-id="" class="android.widget.LinearLayout" package="com.android.keyguard" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="true" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[50,88][430,267]"><node index="0" text="" resource-id="com.android.keyguard:id/clock_view" class="android.widget.RelativeLayout" package="com.android.keyguard" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[50,88][430,238]"><node index="0" text="3:29" resource-id="com.android.keyguard:id/clock_text" class="android.widget.TextView" package="com.android.keyguard" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[137,88][343,238]" /><node index="1" text="PM" resource-id="com.android.keyguard:id/am_pm" class="android.widget.TextView" package="com.android.keyguard" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[355,184][386,213]" /></node><node index="1" text="" resource-id="" class="android.widget.LinearLayout" package="com.android.keyguard" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[50,238][430,267]"><node index="0" text="Tue, September 25" resource-id="com.android.keyguard:id/date_view" class="android.widget.TextView" package="com.android.keyguard" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[157,238][323,267]" /></node></node></node></node><node index="1" text="" resource-id="" class="android.widget.FrameLayout" package="com.android.keyguard" content-desc="null widget." checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[457,76][480,758]"><node NAF="true" index="1" text="" resource-id="" class="android.view.View" package="com.android.keyguard" content-desc="" checkable="false" checked="false" clickable="true" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[469,88][480,746]" /></node></node></node><node index="4" text="" resource-id="com.android.keyguard:id/keyguard_security_container" class="android.widget.FrameLayout" package="com.android.keyguard" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,344][480,854]"><node index="0" text="" resource-id="com.android.keyguard:id/view_flipper" class="android.widget.ViewFlipper" package="com.android.keyguard" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,344][480,854]"><node index="0" text="" resource-id="com.android.keyguard:id/keyguard_selector_view" class="android.widget.FrameLayout" package="com.android.keyguard" content-desc="Slide unlock." checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,356][480,854]"><node index="0" text="" resource-id="" class="android.widget.FrameLayout" package="com.android.keyguard" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,356][480,854]"><node index="0" text="Charged" resource-id="com.android.keyguard:id/keyguard_message_area" class="android.widget.TextView" package="com.android.keyguard" content-desc="" checkable="false" checked="false" clickable="true" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="true" bounds="[0,356][480,381]" /><node index="1" text="" resource-id="com.android.keyguard:id/keyguard_unlock_panel" class="android.widget.FrameLayout" package="com.android.keyguard" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,386][480,854]"><node index="0" text="" resource-id="com.android.keyguard:id/glow_pad_view" class="android.view.View" package="com.android.keyguard" content-desc="Slide area." checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,386][480,854]" /><node index="1" text="" resource-id="com.android.keyguard:id/unread_event_view" class="android.widget.LinearLayout" package="com.android.keyguard" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,386][480,488]" /></node><node index="2" text="" resource-id="com.android.keyguard:id/keyguard_selector_fade_container" class="android.widget.LinearLayout" package="com.android.keyguard" content-desc="" checkable="false" checked="false" clickable="true" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,782][480,854]"><node index="0" text="" resource-id="com.android.keyguard:id/carrier_text" class="android.widget.LinearLayout" package="com.android.keyguard" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,782][480,811]"><node index="0" text="NO SIM CARD — EMERGENCY CALLS ONLY" resource-id="com.android.keyguard:id/carrier" class="android.widget.TextView" package="com.android.keyguard" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="true" bounds="[0,782][480,811]" /></node><node index="1" text="" resource-id="" class="android.widget.LinearLayout" package="com.android.keyguard" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,797][480,854]" /></node><node index="3" text="" resource-id="com.android.keyguard:id/keyguard_selector_view_frame" class="android.view.View" package="com.android.keyguard" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[24,356][456,854]" /></node></node></node></node></node></node></node></hierarchy>
[06:28:58][DEBUG]<airtest.core.android.adb> ....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF shell input touchscreen swipe 240 605 456 605 500 ; echo ---$?---
DEBUG:airtest.core.android.adb:....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF shell input touchscreen swipe 240 605 456 605 500 ; echo ---$?---
[06:28:58][DEBUG]<airtest.core.android.adb> ....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF shell LD_LIBRARY_PATH=/data/local/tmp /data/local/tmp/minicap -i ; echo ---$?---
DEBUG:airtest.core.android.adb:....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF shell LD_LIBRARY_PATH=/data/local/tmp /data/local/tmp/minicap -i ; echo ---$?---
[06:28:58][DEBUG]<airtest.core.android.adb> ....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF shell LD_LIBRARY_PATH=/data/local/tmp /data/local/tmp/minicap -n 'minicap_13652' -P 480x854@480x854/0 -l 2>&1
DEBUG:airtest.core.android.adb:....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF shell LD_LIBRARY_PATH=/data/local/tmp /data/local/tmp/minicap -n 'minicap_13652' -P 480x854@480x854/0 -l 2>&1
[06:28:58][DEBUG]<airtest.utils.nbsp> [minicap_server]b'PID: 4070'
DEBUG:airtest.utils.nbsp:[minicap_server]b'PID: 4070'
[06:28:58][DEBUG]<airtest.utils.nbsp> [minicap_server]b'INFO: Using projection 480x854@480x854/0'
DEBUG:airtest.utils.nbsp:[minicap_server]b'INFO: Using projection 480x854@480x854/0'
[06:28:58][DEBUG]<airtest.utils.nbsp> [minicap_server]b'INFO: (external/MY_minicap/src/minicap_19.cpp:246) Creating SurfaceComposerClient'
DEBUG:airtest.utils.nbsp:[minicap_server]b'INFO: (external/MY_minicap/src/minicap_19.cpp:246) Creating SurfaceComposerClient'
[06:28:58][DEBUG]<airtest.utils.nbsp> [minicap_server]b'INFO: (external/MY_minicap/src/minicap_19.cpp:249) Performing SurfaceComposerClient init check'
DEBUG:airtest.utils.nbsp:[minicap_server]b'INFO: (external/MY_minicap/src/minicap_19.cpp:249) Performing SurfaceComposerClient init check'
[06:28:58][DEBUG]<airtest.utils.nbsp> [minicap_server]b'INFO: (external/MY_minicap/src/minicap_19.cpp:256) Creating virtual display'
DEBUG:airtest.utils.nbsp:[minicap_server]b'INFO: (external/MY_minicap/src/minicap_19.cpp:256) Creating virtual display'
[06:28:58][DEBUG]<airtest.utils.nbsp> [minicap_server]b'INFO: (external/MY_minicap/src/minicap_19.cpp:262) Creating buffer queue'
DEBUG:airtest.utils.nbsp:[minicap_server]b'INFO: (external/MY_minicap/src/minicap_19.cpp:262) Creating buffer queue'
[06:28:58][DEBUG]<airtest.utils.nbsp> [minicap_server]b'INFO: (external/MY_minicap/src/minicap_19.cpp:269) Creating CPU consumer'
DEBUG:airtest.utils.nbsp:[minicap_server]b'INFO: (external/MY_minicap/src/minicap_19.cpp:269) Creating CPU consumer'
[06:28:58][DEBUG]<airtest.utils.nbsp> [minicap_server]b'INFO: (external/MY_minicap/src/minicap_19.cpp:278) Creating frame waiter'
DEBUG:airtest.utils.nbsp:[minicap_server]b'INFO: (external/MY_minicap/src/minicap_19.cpp:278) Creating frame waiter'
[06:28:58][DEBUG]<airtest.utils.nbsp> [minicap_server]b'INFO: (external/MY_minicap/src/minicap_19.cpp:282) Publishing virtual display'
DEBUG:airtest.utils.nbsp:[minicap_server]b'INFO: (external/MY_minicap/src/minicap_19.cpp:282) Publishing virtual display'
[06:28:58][DEBUG]<airtest.utils.nbsp> [minicap_server]b'INFO: (jni/minicap/JpgEncoder.cpp:64) Allocating 1246212 bytes for JPG encoder'
DEBUG:airtest.utils.nbsp:[minicap_server]b'INFO: (jni/minicap/JpgEncoder.cpp:64) Allocating 1246212 bytes for JPG encoder'
[06:28:58][DEBUG]<airtest.utils.nbsp> [minicap_server]b'INFO: (/home/lxn3032/minicap_for_ide/jni/minicap/minicap.cpp:473) Server start'
DEBUG:airtest.utils.nbsp:[minicap_server]b'INFO: (/home/lxn3032/minicap_for_ide/jni/minicap/minicap.cpp:473) Server start'
[06:28:58][DEBUG]<airtest.utils.nbsp> [minicap_server]b'INFO: (/home/lxn3032/minicap_for_ide/jni/minicap/minicap.cpp:475) New client connection'
DEBUG:airtest.utils.nbsp:[minicap_server]b'INFO: (/home/lxn3032/minicap_for_ide/jni/minicap/minicap.cpp:475) New client connection'
[06:28:58][DEBUG]<airtest.core.android.minicap> (1, 24, 4070, 480, 854, 480, 854, 0, 2)
DEBUG:airtest.core.android.minicap:(1, 24, 4070, 480, 854, 480, 854, 0, 2)
it worksit worksit worksit worksit worksit worksit worksit worksit worksit works
[06:29:00][DEBUG]<airtest.core.android.adb> ....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF forward --remove tcp:13652
DEBUG:airtest.core.android.adb:....\venv\lib\site-packages\airtest\core\android\static\adb\windows\adb.exe -s 0123456789ABCDEF forward --remove tcp:13652
Exception in thread Thread-8:
Traceback (most recent call last):
  File "C:\Users\pgi\Anaconda3\lib\threading.py", line 916, in _bootstrap_inner
    self.run()
  File "C:\Users\pgi\Anaconda3\lib\threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "C:/Users/pgi/Downloads/adbmirror/adbmirror/gui/filebar/pr.py", line 39, in feedframe
    for stream in self.device.minicap.get_stream():
  File "....\venv\lib\site-packages\airtest\core\android\minicap.py", line 245, in _get_stream
    header = s.recv(4)
  File "....\venv\lib\site-packages\airtest\utils\safesocket.py", line 32, in recv
    raise socket.error("socket connection broken")
OSError: socket connection broken

Process finished with exit code -1
```

**Screenshots**

```py
import concurrent.futures
import tkinter as tk
from lxml import etree
from io import StringIO
from airtest.core import android as andr
from multiprocessing.dummy import Pool
from threading import Thread
import multiprocessing
import cv2
import queue
import re
import numpy as np
from PIL import Image
import logging

logging.basicConfig(level=logging.DEBUG)


class HandleScreen():
    def __init__(self):
        self.adb = andr.android.ADB()
        self.device = andr.android.Android()
        self.best_ration = self.calc_aspect3()
        self.the_q = queue.Queue(1)
        self.stopped_show = False


    def calc_aspect3(self):
        device_data = self.device.display_info
        self.root = tk.Tk()

        screen = (self.root.winfo_screenwidth(),self.root.winfo_screenheight())
        screen_android = (device_data["width"], device_data["height"])

        max_h = min(screen[1], screen_android[1])
        max_w = min(screen[0], screen_android[0])
        best_ration = min(max_w / screen_android[0], max_h / screen_android[1])

        if screen[0] > screen_android[0] and screen[1] > screen_android[1]:
            return 1
        return best_ration


    def feedframe(self):
        frame_reader = self.device.minicap
        while True:
            frame = frame_reader.get_frame_from_stream()
            self.the_q.put(frame)

    def prepareandstartscreen(self):
        Thread(target=self.feedframe).start()
        Thread(target=self.show_screen).start()
        return self

    def RUN(self):
        self.prepareandstartscreen()
        if self.device.is_locked():
            self.device.unlock()
            # self.device.keyevent("POWER")
            # Thread(target=self.unlock_screen()).start()
        print("After")


    def show_screen(self):
        cv2.namedWindow("resized frame")
        while not self.stopped_show:
            stream = self.the_q.get()
            img = cv2.imdecode(np.frombuffer(stream, np.uint8), -1)
            small = cv2.resize(img, (0, 0), fx=0.7, fy=0.7)
            cv2.imshow("resized frame", small)
            key = cv2.waitKey(1) & 0xFF
            if key == 27:
                self.stopped_show = True

    def extract_metadata(self):
        parser = etree.XMLParser(recover=True, encoding="utf-8")
        stream = self.device.shell("uiautomator dump  && cat /sdcard/window_dump.xml").\
            split("<?xml version="1.0" encoding="UTF-8" standalone="yes" ?>")[1]
        print(stream)
        xml = etree.parse(StringIO(stream), parser)
        return xml

    def search_icon_pos(self, xml):
        """return boundary string from xml"""
        logging.info("searching for search icon pos")
        search_icon = xml.xpath(
            "//*[@class="android.widget.ImageView" and @index="0"]|//*[@class="android.widget.EditText" and @text="Search"]")
        boundary = None
        if search_icon:
            search_icon = search_icon[-1].get("bounds")
            boundary = list(map(int, re.findall("\d+", search_icon)))
        return boundary

    def extract_result_positions(self, xml, window_type: str):
        action_dict = {
            "RecyclerView": "//*[@class="org.telegram.messenger.support.widget.RecyclerView"]/node",
            "group_header": "//*[@class="org.telegram.messenger.support.widget.RecyclerView"]/preceding-sibling::node/node[2]",
            "add_tag_or_invite": "//*[@class="org.telegram.messenger.support.widget.RecyclerView"]/node[@class="android.widget.FrameLayout" and count(node)=1]",
            "member": "//*[@class="org.telegram.messenger.support.widget.RecyclerView"]/node[@class="android.view.View"]"
        }
        if window_type in action_dict:
            content_meniu = action_dict[window_type]
            content_meniu = xml.xpath(content_meniu)
            nodes = []
            for node in content_meniu:
                search_icon = node.get("bounds")
                boundary = list(map(int, re.findall("\d+", search_icon)))
                yield boundary
        else:
            raise KeyError()

    def unlock_screen(self):
        xml = self.extract_metadata()
        lockWindowBounds = xml.xpath("//*[@resource-id="com.android.keyguard:id/keyguard_selector_view_frame"]")[0].get("bounds")
        lockWindowBounds = boundary = list(map(int, re.findall("\d+", lockWindowBounds)))

        data = dict(zip(["left", "bottom","right","top"], lockWindowBounds))
        startPoint_x = (data["right"] + data["left"]) / 2
        startPoint_y = (data["bottom"] + data["top"]) / 2
        self.device.swipe((startPoint_x, startPoint_y), (data["right"], startPoint_y))


    def do_actions(self):
        xml = self.extract_metadata()
        pos = self.search_icon_pos(xml)
        search_pos = pos if pos else None
        self.device.shell("input tap {} {}".format(search_pos[0], search_pos[1]))
        self.device.shell("input text {}".format("botsgroup"))
        xml = self.extract_metadata()

        result = next(self.extract_result_positions(xml, "RecyclerView"))
        self.device.shell("input tap {} {}".format(result[0], result[1]))


screen = HandleScreen()
screen.RUN()`
```

**Smartphone (please complete the following information):**
 - Device:  lenovo s660, s850 plus (Chinese fake)
 - OS: kitkat

I found the work around by using get_frame()
Basically I am trying to design a tool which while mirror screen it does some action such that u can see all of them in the same screen

**Comment by yimelia** [comment]
hi @YurieCo , I try your script and it seems works well, I can see the device screen. try to upgrade your airtest?
